---
title: "op19 - Re-quantification of ATAC-seq"
subtitle: "Differential accessibility analysis"
author: "Stephanie Strohbuecker"
output:
  html_document:
    df_print: paged
---

```{r, eval=FALSE, include=FALSE}
rmarkdown::render("/camp/stp/babs/working/strohbs/projects/guillemotf/oana.paun/op17_PM21134_ChIP_SWI_SNFxASCL1/vignettes/4-5.op19_DiffBind_analysis_with_HP_regions_ASCL1KO_def.Rmd", output_file = "/camp/stp/babs/working/strohbs/projects/guillemotf/oana.paun/op17_PM21134_ChIP_SWI_SNFxASCL1/vignettes/4-5.op19_DiffBind_analysis_with_HP_regions_ASCL1KO_def.html")
```

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      eval.after = "fig.cap",
                      fig.width = 12)
```

Investigator: Oana Paun
PI:Francois Guillemot

# Project summary

(taken from the project proposal form)
    
"
An initial ATAC-seq experiment (Lims ID PM19276, run  200120_K00102_0447_AHFVFTBBXY) was carried out and Harshil Patel performed the peak calling and differential accessibility analysis. From this experiment, we were particularly interested in the following comparisons:
1) NP_WT vs NRS_WT
2) NRS_ASCL1_KO vs NRS_WT

A second ATAC-seq experiment  (Lims ID PM21132, run  210723_A01366_0037_BH2NLTDMXY) was carried out and Miriam Llorian Sopena performed the peak calling and differential accessibility analysis. From this experiment, we were particularly interested in the following comparisons:
1) 48h_DMSO vs WT
2) 48h_BRM014 vs 48h_DMSO
3) 48h_BRM014 vs WT

While some of the samples were unique to each of the two experiments, the WT samples come from the same cells (human iPSC-derived neuronal cells at day 24 post neural induction). However, we noticed approximately 55,000 accessible regions in the consensus set of peaks in WT cells from experiment 1, and only approximately 33,000 accessible regions in the consensus set of peaks in WT cells from experiment 2. Since ultimately we are interested in the regions which loose accessibility in both ASCL1_KO and BMR014 mutants relative to the WT cells, we find this discrepancy in their relative WT controls unexpected. Having a closer look at specific genomic regions, we concluded that, most likely, this discrepancy comes from the different signal to noise ratios in the two experiments. More specifically, two of the three wild-type replicates we provided as wild-type samples in the second experiment seem to have a lower signal to noise ration when compared to the third one from the same experiment, or with the other two wild-type replicates from the first experiment. 

Therefore, we would like to generate a consensus set of peaks from the two experiments and use them for the downstream differential accessibility analysis. To be able to carry on the analysis described above, I swiched to “—min_reps_consensus 1” for both experiments (since most likely the discrepancy between the two experiments came from the weak signal to noise ratio in the two replicates from the second experiment). I then overlapped the wild-type peaks from both experiments (1bp overlap) and generate what we will now call the consensus set of peaks for the wild-type samples (called “wt_nrs_consensus_peaks.txt”).

After further discussions with Miram Llorian Sopena, I also used the same criteria (“—min_reps_consensus 1”) to generate new consensus sets of peaks for all the other genotypes. 

Using the provided sets of consensus peaks and specifically only for the peaks identified in the wt neurons (“wt_nrs_consensus_peaks.txt”), I would like to ask for differential accessibility analysis for the following comparisons: 

1) –  wt progenitors (“wt_np_consensus_peaks.txt”) vs  wt neurons (“wt_nrs_consensus_peaks.txt”)
– ascl1ko neurons (“ascl1ko_nrs_consensus_peaks.txt)  vs  wt neurons (“wt_nrs_consensus_peaks.txt”)

Please find the IDs corresponding to the raw data of these samples highlighted in the  attached PM19276 Excel spreadsheet.

2) – 48h dmso neurons (“48h_dmso_nrs_consensus_peaks.txt”) vs wt neurons (“wt_nrs_consensus_peaks.txt”)
– 48h brm014 neurons (“48h_brm014_nrs_consensus_peaks.txt”) vs wt neurons (“wt_nrs_consensus_peaks.txt”)
              – 48h brm014 neurons   
              (“48h_brm014_nrs_consensus_peaks.txt”) vs 48h dmso 
              neurons (“48h_dmso_nrs_consensus_peaks.txt”)

Please find the IDs corresponding to the raw data of these samples highlighted in the  attached PM21132 Excel spreadsheet.
"

In this iteration we use the bam files generated previously with the consensus peak set identified by Harshil Patel (PM19276).

```{r}
RMD_file_name <- "4-5.op19_DiffBind_analysis_with_HP_regions_ASCL1KO_def"
```



```{r message=FALSE, warning=FALSE}
## Libraries
library(tidyverse)
library(here)

library(ggpubr)

library(plyranges)
library(DiffBind)
```

```{r message=FALSE, warning=FALSE}
# load_all() makes the functions in the R/ directory available for usage
pkgload::load_all()
```

```{r message=FALSE, warning=FALSE}
# Path to output and figures directories
bam_files <-
  list.files(
    "/camp/stp/babs/outputs/guillemotf/oana.paun/stephanie.strohbuecker/op17_PM21134_ChIP_SWI_SNFxASCL1/ATACseq_re-analysis",
    pattern = ".bam$",
    recursive = TRUE,
    full.names = TRUE
  ) %>% 
  set_names(str_remove(basename(.), ".mLb.clN.sorted.bam")) %>%
  enframe(name = "SampleID", "bamReads") %>% 
  mutate(project_lims = basename(dirname(dirname(bamReads))),
         SampleID = case_when(
           project_lims == "PM21132" ~ SampleID,
           TRUE ~ str_replace(SampleID, "_R", "_PM19276_R") %>% 
             str_replace("NRS_ASCL1", "ASCL1")
         ))
  
peak_files <-
   "/camp/stp/babs/outputs/guillemotf/oana.paun/stephanie.strohbuecker/op17_PM21134_ChIP_SWI_SNFxASCL1/ATACseq_re-analysis/PM19276/consensus_peaks/consensus_peaks.mLb.clN.bed"

macs_peak_files <-
list.files(
    "/camp/stp/babs/outputs/guillemotf/oana.paun/stephanie.strohbuecker/op17_PM21134_ChIP_SWI_SNFxASCL1/ATACseq_re-analysis",
    pattern = ".xls",
    full.names = TRUE,
    recursive = TRUE
  ) %>%
  set_names(str_remove(basename(.), ".mLb.clN_peaks.xls")) %>%
  enframe(name = "SampleID", "Peaks")%>% 
  mutate(project_lims = basename(dirname(dirname(Peaks))),
         SampleID = case_when(
           project_lims == "PM21132" ~ str_replace(SampleID, "_R", "_PM21132_R") %>% 
             str_remove("48h_"),
           TRUE ~ str_replace(SampleID, "_R", "_PM19276_R") %>% 
             str_replace("NRS_ASCL1", "ASCL1")
         )) 

sample_info <-
  read_tsv(
    here(
      "data",
      "derived_data",
      "Tables",
      "4-1.Re-quantify_ATAC_op19",
      "ATAC_seq_op19_sample_information.tsv"
    )
  ) %>%
  select(Sample_limsid,
         group,
         replicate,
         Sample_Genotype,
         project_lims,
         Sample_Name) %>%
  distinct() %>%
  mutate(
    SampleID_display = str_c(project_lims, "_R", replicate, "_", Sample_Genotype, sep = ""),
    SampleID = str_c(group, "_R", replicate, sep = ""),
    Sample_Genotype = str_replace_all(Sample_Genotype, " ", "_"),
    PeakCaller = "macs",
    Treatment = case_when(str_detect(group, "NP") ~ "NP",
                          TRUE ~ "NRS")
  ) %>%
  left_join(bam_files, by = c("project_lims", "SampleID")) %>%
  left_join(macs_peak_files, by = c("project_lims", "SampleID")) %>%
  select(-SampleID) %>%
  dplyr::rename(SampleID = SampleID_display) %>%
  group_by(project_lims) %>%
  group_split() %>%
  set_names(map(., ~ unique(.x$project_lims)))

output_dirs <-
  str_c(
    here(),
    "data/derived_data",
    c("Tables", "R_objects", "Figures"),
    RMD_file_name,
    sep = "/"
  ) %>%
  set_names(c("Tables", "R_objects", "Figures"))

walk(output_dirs, ~
       generate_directories(.x))
```

```{r message=FALSE, warning=FALSE}
consensus_peaks <- read_tsv(peak_files,
         col_names = c("seqnames", "start", "end", "interval_id", "score", "strand")) %>%
  select(-starts_with("X")) %>% 
  as_granges()

exp_gr <- list(
  "PM19276" =consensus_peaks,
  "PM21132" =  consensus_peaks %>% 
  as_tibble() %>% 
  mutate(seqnames = str_remove(seqnames, "chr")) %>% 
  as_granges()
)
```


```{r}
blacklist <- read_bed(
  here(
    "data",
    "derived_data",
    "Tables",
    "3-2.ABC_algorithm_TSS_annotation",
    "ENCFF001TDO_blocklisted_regions_GRCh37.ensembl.sort.bed"
  )
)
```

```{r}
sample_info <- sample_info %>%
  map(
    ~ .x %>%  select(
      SampleID,
      Tissue = project_lims,
      Factor = group,
      Condition = Sample_Genotype,
      Treatment,
      Replicate = replicate,
      bamReads,
      Peaks,
      PeakCaller
    )
  )

```

For the analysis of differential accessibility we used the Bioconductor package [DiffBind](https://bioconductor.org/packages/release/bioc/html/DiffBind.html) (version `r packageVersion("DiffBind")`). `DiffBind` is designed to identify genomic regions that are differentially bound (in case of ChIP-seq data analysis) or accessible (in case of ATAC-seq data analysis). The package contains functions to support the processing of peak sets, counting sequencing reads overlapping intervals in peak sets, and identifying statistically significantly differentially used sited (measured by differences in read densities). The used underlying statistical approach is [DESEq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), originally developed for RNA-seq data analysis.

```{r include=FALSE}
dba_ls <- map(sample_info, function(sample_sheet){
  dba_obj <- dba(sampleSheet = sample_sheet,
    minOverlap = 0)
  dba_obj
})

```

```{r}
# Apply blacklist
# dba_ls <- map(dba_ls, function(dba_obj) {
#   dba_obj <-
#     dba.blacklist(dba_obj, blacklist = blacklist, greylist = blacklist)
#   dba_obj
# })
```

```{r}
dba_ls <- map(dba_ls, function(dba_obj) {
  dba_obj$config$doGreylist  <- FALSE
  dba_obj$config$doBlacklist <- FALSE
  dba_obj$config$cores <- 6
  dba_obj
})
```

Counting is performed against the consensus peak defined by Harshil Patel using 2 replicates (i.e based on project PM19276). 
```{r include=FALSE}
dba_ls <- imap(dba_ls, function(dba_obj, exper) {
  dba_obj <-
    dba.count(dba_obj,
              peaks = exp_gr[[exper]],
              summits = FALSE,
              filter = 0
              )
})
```

```{r}
# write_rds(dba_ls, file = str_c(output_dirs[["R_objects"]], "dba_count_object.RDS", sep = "/"), compress = "gz")
```

```{r}
# dba_ls <- read_rds(file = str_c(output_dirs[["R_objects"]], "dba_count_object.RDS", sep = "/"))
```

As enrichment in ATAC-seq is expected to occur over relatively narrow intervals (roughly 100-600bp), it is expected that there should not be systematic differences in signals over much larger intervals (on the order of 10000bp and greater). Any differences seen should be technical rather than biological, so we can normalise counts based on these differences. This background normalisation is recommended for ATAC-seq analysis ([DiffBind vignette](https://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf)).

```{r include=FALSE}
# Normalise data
dba_ls <- map(dba_ls, function(dba_obj) {
  dba_obj <-
    dba.normalize(
      dba_obj,
      background = TRUE#,
      # normalize = DBA_NORM_NATIVE, this approach increases the unexpected change in direction
      # method = DBA_DESEQ2
    )
})
```

After normalisation we can use the normalised read counts to investigate the clustering of the samples using cross-correlations.

```{r}
# Plot sample relationships after normalisation
iwalk(dba_ls, function(dba_obj, dba_name) {
  plot(dba_obj, main = dba_name)
})
```

The different sample conditions in PM19276 separate better from each other than the sample conditions found in experiment PM21132.

The information about the peaks and their normalised counts in each of the samples is provided in a tab-delimited file for download at the end of this report.
```{r}
# generate table of normalised abundancies with
imap(dba_ls, function(dba_obj, dba_name) {
  left_join(
    dba.peakset(dba_obj, bRetrieve = TRUE) %>%
      as_tibble() %>%
      select(-strand),
    exp_gr[[dba_name]] %>% as_tibble(),
    by = c("seqnames", "start", "end", "width")
  ) %>%
    mutate(seqnames = as.character(seqnames),
      seqnames = case_when(
      str_detect(seqnames, "chr") ~ seqnames,
      TRUE ~ str_c("chr", seqnames, sep = "")
    )) %>% 
    relocate(strand, .after = width) %>%
    relocate(interval_id, .after = strand) %>%
    write_tsv(
      str_c(
        output_dirs[["Tables"]],
        "/",
        dba_name,
        "_consensus_peak_set_normalised_counts_DiffBind.tsv"
      )
    )
})
```

# Perform differential accessibility analysis

After normalising the experimental data, we used performed pairwise comparisons between the available genotypes (conditions) for each of the experiments using the model design: `~ Condition`

```{r include=FALSE}
# set contrasts
dba_ls <- map(dba_ls, function(dba_obj) {
  dba_obj <-
    dba.contrast(dba_obj, categories = DBA_CONDITION,
                 minMembers = 2)
})
```


```{r}
# Ensure that the correct contrasts were chosen
# map(dba_ls, function(dba_obj) {
#     dba.show(dba_obj, bContrasts = TRUE)
# })
```

Using DESeq2 the differential accessibility analysis is carried out. This will assign a p-value and FDR to each of the peak regions of interest indicating confidence that they are differentially accessible. Significantly differentially accessible peaks are identified using the default threshold of FDR <= 0.05.

```{r include=FALSE}
dba_ls <- map(dba_ls, function(dba_obj) {
  dba_obj <-
    dba.analyze(dba_obj, bParallel = FALSE, method=DBA_DESEQ2, design = ~ Condition)#, bRetrieveAnalysis = TRUE)
})
```

```{r}
## The warning that dba.analyze generated "system is singular" let to the investigation of the dispersion estiamtes. They look resonable, we therefore ignored the warning.
# plotDispEsts(dba_ls_DESeq2$PM19276)
```

The number of differentially accessible peak regions is shown below (in the column titled `DB.DESeq2`).
```{r}
imap(dba_ls, function(dba_obj, dba_name) {
    dba.show(dba_obj, bContrasts = TRUE) %>% 
    as_tibble() %>%
    mutate(LIMS_ID = dba_name)
}) %>% 
  purrr::reduce(bind_rows)
```

No peak regions are significantly differentially accessible between WT and DMSO-treated WT samples in experiment PM21132. This is in line with the expectation that the vehicle treatment with DMSO does not alter chromatin accessibility. The number of significantly differentially accessible peaks is generally lower in the pairwise comparisons for experiment PM21132 than in experiment PM19276.

Oana Paun noticed an unexpected direction of change when comparing ASCL1 KO with the wildtype conditions. We visualised all significantly differential accessible peaks in a MA plot. MA plots are a useful way to visualize the relationship between the overall accessibility level at each site and the magnitude of the change between conditions, as well as the effect of normalization on data.
```{r}
imap(dba_ls, function(dba_obj, dba_name) {
  contrs <- dba.show(dba_obj, bContrasts = TRUE) %>%
    as_tibble() %>%
    filter(DB.DESeq2 > 0)
  map(1:NROW(contrs), ~ {
    dba.plotMA(dba_obj, contrast = .x) 
  })
})
```

```{r}
# imap(dba_ls, function(dba_obj, dba_name) {
#   contrs <- dba.show(dba_obj, bContrasts = TRUE) %>%
#     as_tibble() %>%
#     filter(DB.DESeq2 > 0)
#   map(1:NROW(contrs), ~ {
#     dba.plotMA(dba_obj, contrast = .x, bNormalized = FALSE)
#   })
# })
```

```{r}
dba_res <- imap(dba_ls, function(dba_obj, dba_name) {
  avail_contr <-  dba.show(dba_obj, bContrasts = TRUE) %>%
    as_tibble() %>%
    mutate(comparison = str_c(Group, "vs", Group2, sep = "_"))
  map(1:NROW(avail_contr), function(contr_nb) {
    current_contr <- avail_contr[contr_nb,]
    if (current_contr$DB.DESeq2 != 0) {
      # with read counts adjusted by the normalisation factors for calculating concentration values; fold changes are reported using the potentially shrunk values
      #concentration - mean (log) reads across all samples in both groups
      # Fold – mean fold difference of binding affinity of group 1 over group 2 (Conc1 - Conc2). Absolute value indicates magnitude of the difference, and sign indicates which one is bound with higher affinity, with a positive value indicating higher affinity in the first group
      dba.report(
        dba_obj,
        contrast = contr_nb,
        th = 0.05,
        bNormalized = TRUE
      ) %>%
        as_tibble() %>%
        mutate(contrast = current_contr$comparison) %>%
        as_granges()
    }
  }) %>%
    set_names(map(., ~ unique(.$contrast))) %>%
    compact()
}) 
```

The information about significantly differential accessible peak regions are available in `.xlsx` format for download at the end of this report.
In addition to the peak information (chromosome, start, end and strand), the columns show the mean read concentration over all the samples (`Conc`) and the mean concentration over the samples in each of the first and second group. The `Fold` column shows the log fold changes between the two groups as calculated by the DESeq2 analysis. A positive value indicates increased accessibility in the first group compared to the second group. Two further columns give confidence measures for identifying the peaks as differentially accessible, with a raw p-value and a multiple-testing corrected FDR (also calculated by the DESeq2 analysis).
```{r message=FALSE, warning=FALSE, include=FALSE}
iwalk(dba_res, function(dba_res_obj, res_name) {
  openxlsx::write.xlsx(dba_res_obj, str_c(
    output_dirs[["Tables"]],
    str_c(
      res_name,
      "differentially_accessible_peaks_ATACseq.xlsx",
      sep = "_"
    ),
    sep = "/"
  ), overwrite = TRUE)
})
```

```{r}
dba_all <- imap(dba_ls, function(dba_obj, dba_name) {
  avail_contr <-  dba.show(dba_obj, bContrasts = TRUE) %>%
    as_tibble() %>%
    mutate(comparison = str_c(Group, "vs", Group2, sep = "_"))
  map(1:NROW(avail_contr), function(contr_nb) {
    current_contr <- avail_contr[contr_nb,]
    if (current_contr$DB.DESeq2 != 0) {
      # with read counts adjusted by the normalisation factors for calculating concentration values; fold changes are reported using the potentially shrunk values
      #concentration - mean (log) reads across all samples in both groups
      # Fold – mean fold difference of binding affinity of group 1 over group 2 (Conc1 - Conc2). Absolute value indicates magnitude of the difference, and sign indicates which one is bound with higher affinity, with a positive value indicating higher affinity in the first group
      dba.report(
        dba_obj,
        contrast = contr_nb,
        th = 1,
        bNormalized = TRUE
      ) %>%
        as_tibble() %>%
        mutate(contrast = current_contr$comparison) %>%
        as_granges()
    }
  }) %>%
    set_names(map(., ~ unique(.$contrast))) %>%
    compact()
}) 
```

```{r message=FALSE, warning=FALSE, include=FALSE}
iwalk(dba_all, function(dba_res_obj, res_name) {
  openxlsx::write.xlsx(dba_res_obj, str_c(
    output_dirs[["Tables"]],
    str_c(
      res_name,
      "all_accessible_peaks_ATACseq.xlsx",
      sep = "_"
    ),
    sep = "/"
  ), overwrite = TRUE)
})
```

```{r message=FALSE, warning=FALSE}
dba_res_comb <-
  imap(dba_ls, function(dba_obj, dba_name) {
    avail_contr <-  dba.show(dba_obj, bContrasts = TRUE) %>%
      as_tibble(rownames = "contrast") %>%
      mutate(comparison = str_c(Group, "vs", Group2, sep = "_"))
    contrs <- avail_contr %>%
      filter(DB.DESeq2 > 0) %>%
      mutate(contrast = as.numeric(contrast)) %>%
      pull(contrast)
    
    dba.report(
      dba_obj,
      contrast = contrs,
      th = 0.05,
      bNormalized = TRUE,
      bDB = TRUE
    )
  })
```
We can visualise the overlap of significantly differentially peaks in the different comparisons within each experiment. In experiment PM19276, for example, 5520 peaks are differentially accessible both in the comparison between ASCL1_KO and NP_WT as well as between ASCL1_KO and NRS_WT.
```{r, fig.height=6}
iwalk(dba_res_comb, function(repObj, dba_name) {
  overlaps <-
    dba.plotVenn(repObj,
                 1:length(repObj$peaks),
                 main = str_c(dba_name, ": Peak Site Overlaps", sep = ""))
})
```

The significantly differentially accessible peak regions can be divided into those that become more accessible and those that loose accessibility in the first condition in comparison to the second condition.

```{r}
imap_dfr(dba_res, function(dba_res_obj, dba_name) {
  imap_dfr(dba_res_obj, function(peak_ls, comp_name) {
    sites_of_int <- peak_ls %>%
      arrange(-Fold) %>%
      mutate(direct = case_when(Fold > 0 ~ "Gain",
                                Fold < 0 ~ "Loss")) %>%
      
      as_tibble() %>%
      group_by(direct) %>%
      summarise(nb_peaks = n()) %>%
      mutate(experiment = dba_name,
             contrast = comp_name) %>% 
      select(experiment, contrast, direction = direct, nb_peaks)
    sites_of_int
  })
}) %>% 
  DT::datatable(caption = "Number of peaks gained or lost in the indicated comparison")
```


```{r}
# # Oana Paun noticed that, against expectations, the number of more accessible peaks seems to increase in the ASCL1-KO in comparison to wildtype conditions (in experiment PM19276). We investigated if there is a difference in the effect size in those regions that gain accessibility in ASCL1-KO conditions.
# imap(dba_res, function(dba_res_obj, dba_name) {
#   imap(dba_res_obj, function(peak_ls, comp_name) {
#    peak_ls %>% 
#       as_tibble() %>%
#       mutate(direct = case_when(Fold > 0 ~ "Gain",
#                                 Fold < 0 ~ "Loss")) %>% 
#       # filter(abs(Fold) > 0.5) %>%
#       ggplot(aes(abs(Fold), colour = direct)) +
#       geom_density() +
#       ggtitle(comp_name)
#   }) %>% 
#     ggarrange(plotlist = .) %>% 
#     annotate_figure(top = dba_name)
# }) 
```


```{r}
## Visualising the absolute LFC of the provided peak regions we can identify a proportion of peaks that show gain accessibility but at a smaller effect size (< 0.5). This indicates that if a filter is applied to the effect size, a larger proportion of peak regions that gain accessibility is filtered out (a smaller proportion of peaks that loose accessibility is filtered out). Nonetheless as there is a considerably higher number of peaks that gain accessibility, there will still be a higher number of peaks that gain accessibility.
# imap_dfr(dba_res, function(dba_res_obj, dba_name) {
#   imap_dfr(dba_res_obj, function(peak_ls, comp_name) {
#     sites_of_int <- peak_ls %>%
#       arrange(-Fold) %>%
#       mutate(direct = case_when(Fold > 0 ~ "Gain",
#                                 Fold < 0 ~ "Loss")) %>%
#       
#       as_tibble() %>%
#       filter(abs(Fold) > 0.5) %>% 
#       group_by(direct) %>%
#       summarise(nb_peaks = n()) %>%
#       mutate(experiment = dba_name,
#              contrast = comp_name) %>% 
#       select(experiment, contrast, direction = direct, nb_peaks)
#     sites_of_int
#   })
# }) %>% 
#   DT::datatable(caption = "Number of peaks gained or lost in the indicated comparison")
```

As we are using a custom peak set for the analysis, the analysis may be skewed as these peaks may not be detected across all samples.

DiffBind then calculates linear scaling factors from either the total number of reads in each library, which assumes that true global differences may be expected and technical bias is small, or the total number of reads in queried peak regions, which should eliminate global differences in favor of reducing any technical biases. 


For the visualisation of peak profiles, we selected peaks with at least a LFC (Fold) of 1 (fold change of 2).
```{r}
sites_of_interest <- imap(dba_res, function(dba_res_obj, dba_name) {
  imap(dba_res_obj, function(peak_ls, comp_name) {
    sites_of_int <- peak_ls %>%
      filter(abs(Fold) > 1) %>%
      arrange(-Fold) %>%
      mutate(direct = case_when(Fold > 0 ~ "Gain",
                                Fold < 0 ~ "Loss")) %>%
      group_by(direct) %>%
      group_split()
    names(sites_of_int)[1] <- unique(sites_of_int[[1]]$direct)
    names(sites_of_int)[2] <- unique(sites_of_int[[2]]$direct)
    sites_of_int
  })
})
```

The profile plots below show a maximum of 1000 sites, i.e. if there are more than a 1000 peaks (with an absolut LFC greater 1 due to the selection above), only the 1000 sites with the greatest absolute value fold change will be included.

The profiles below show the relative intensity of overlapping peaks in each bin for each peak in each sample, along with summary plots showing the average profile across the sites for each sample. Sample groups will be merged such that each sample class will have one heatmap based on the normalised mean read counts for all the samples in that class.

```{r}
imap(sites_of_interest, function(exp_sites, exp_name) {
  imap(exp_sites, function(peak_set, comp_name) {
    sample_names <- str_split(comp_name, "_vs_") %>% unlist()
    sample_select <- list(dba_ls[[exp_name]]$mask[[sample_names[1]]],
                          dba_ls[[exp_name]]$mask[[sample_names[2]]]) %>%
      set_names(sample_names)
    profiles <-
      dba.plotProfile(
        dba_ls[[exp_name]],
        sites = peak_set,
        samples = sample_select
    )
    dba.plotProfile(profiles)
  })
})
```

We can also visualise the top 1000 significantly differential sites in a heatmap (without pre-selection before plotting).
```{r, fig.height=6}
readscores <- imap(dba_ls, function(dba_obj, dba_name) {
  avail_contr <-  dba.show(dba_obj, bContrasts = TRUE) %>%
    as_tibble() %>%
    mutate(comparison = str_c(Group, "vs", Group2, sep = "_"))
  map(1:NROW(avail_contr), function(contr_nb) {
    current_contr <- avail_contr[contr_nb, ]
    if (current_contr$DB.DESeq2 != 0) {
      readscores <- dba.plotHeatmap(
        dba_obj,
        contrast = contr_nb,
        correlations = FALSE,
        colScheme = colorRampPalette(c("red", "black", "green"))(n = 13),
        main = str_replace_all(current_contr$comparison, "_", " ")
      )
      readscores
    }
  }) %>%
    set_names(map(., ~ unique(.$contrast))) %>%
    compact()
}) 
```

Information about the used peak set and the significantly differentially accessible peaks and all peaks considered for the differentially expression analysis can be downloaded using the links below.

```{r}
excel_files_path <-
  list.files(output_dirs["Tables"], full.names = TRUE) %>%
  set_names(basename(.))
```

```{r, results="asis"}
xfun::pkg_load2(c('base64enc', 'htmltools', 'mime'))

imap(excel_files_path, function(file_path, file_text) {
  xfun::embed_file(file_path,
                   text = file_text)
})
```



