---
title: "PM21134 - ChIP of SWI/SNF and ASCL1"
author: "Stephanie Strohbuecker"
subtitle: Read alignment using the nf-core RNA-seq pipeline for Y-chromosome removal
output:
  html_document:
    df_print: paged
---

Re-alignment for the RNA-seq (analysed by Harshil Patel) to remove y-chromosome mapping reads before making the data publicly available:

* RNA-seq: RN19275
    - run: 200128_K00102_0451_BHGHN3BBXY

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      eval.after = "fig.cap",
                      fig.width = 12)
```

We are using the [nf-core RNAseq pipeline (version 3.7)](https://nf-co.re/rnaseq/3.7) for aligning the RNA-seq reads.


```{r}
RMD_file_name <- "6-1.RN19275_nfcore_RNA-seq"
```

## Libraries

```{r}
library(tidyverse)
library(here)
```

```{r}
pkgload::load_all() # load all the functions available in this project
```

```{r}
log_dir <- here("analysis", "logs")
scripts_dir <- here("analysis", "scripts")
```

```{bash, eval=FALSE}
cp /camp/stp/babs/inputs/sequencing/data/guillemotf/oana.paun/RN19275/RN19275_design.csv ../data/raw_data/project_docs/RNAseq_RN19275_design.csv
```


## RNA-seq

```{r}
sample_info <-
  readxl::read_excel(here(
    "data",
    "raw_data",
    "project_docs",
    "samples_for_data_upload_vs12-08-2022.xlsx"
  )) %>% 
  select(project_id, sample_id, sequencing_type, sequencing_mode, flowcell) %>% 
  filter(project_id == "RN19275")

RNA_sample_info <-
  read_csv(here(
    "data",
    "raw_data",
    "project_docs",
    "RNAseq_RN19275_design.csv"
  )) %>%
  filter(sample %in% sample_info$sample_id) %>% 
  select(sample, fastq_1 = file) %>% 
  mutate(fastq_2 = NA_character_,
         strandedness = "reverse")
```

Note that the fastq files had to be retrieved from archive. They were temporarily stored here:
`/camp/stp/babs/scratch/strohbs/projects/guillemotf/oana.paun/RN19275_200128_K00102_0451_BHGHN3BBXY/raw_data`

```{r}
RNA_sample_info <- RNA_sample_info %>%
  mutate(
    fastq_1 = str_c(
      "/camp/stp/babs/scratch/strohbs/projects/guillemotf/oana.paun/RN19275_200128_K00102_0451_BHGHN3BBXY/raw_data",
      basename(fastq_1),
      sep = "/"
    )
  ) 
```


To run read alignment using the RNA nf-core pipeline, we need to prepare a samplesheet with the following columns:

* sample
* fastq read 1
* fastq read 2 
* strandedness

```{r}
generate_directories(str_c(here(), "data", "derived_data", "nfcore_RNA_RN19275", sep = "/"))
```

```{r}
RNA_sample_info %>%
  write_csv(
    file = str_c(
      here(),
      "data",
      "derived_data",
      "nfcore_RNA_RN19275",
      "design.csv",
      sep = "/"
    ),
    na = ""
  )
```

The results for the nf-core RNA-seq pipeline will be stored here:
```{r}
str_c(here(), "data", "derived_data", "nfcore_RNA_RN19275", sep = "/")
```

The `custom.config` is empty 
```{r}
write_lines(
  "",
  file = str_c(
    here(),
    "data",
    "derived_data",
    "nfcore_RNA_RN19275",
    "custom.config",
    sep = "/"
  )
)
```

(Note that the slurm command needs to be run from the following directory: `/camp/stp/babs/working/strohbs/projects/guillemotf/oana.paun/op17_PM21134_ChIP_SWI_SNFxASCL1/data/derived_data/nfcore_RNA_RN19275`).

```{r}
slurm_cmd <-
  str_c(
    "sbatch -c 1 --mem-per-cpu 7G -N 1 --time 16:00:00 --output",
    str_c(log_dir, "nfcore_RNA_RN19275.log", sep = "/"),
    "--error",
    str_c(log_dir, "nfcore_RNA_RN19275_error.log", sep = "/"),
    str_c(scripts_dir, "6-1a.Run_nfcore_RNA_RN19275.sh", sep = "/"),
    sep = " "
  ) 

slurm_cmd
```

